Internal cross-references MEP

rowanc1 opened the pull request:
# 📖 [Read the MEP Here](https://myst-eps--10.org.readthedocs.build/en/10/meps/mep-0002/)

📆 MEP will close Monday March 13th, 2023, PT 📆

The current draft with most comments/todos resolved, the [Companion issue](https://github.com/executablebooks/myst-enhancement-proposals/issues/9#user-content-fn-1-f569d755bac79a13c45d421e4f925398) for the MEP can host overall discussions.

**Outstanding issues / questions:**
- [x] ~agree on using three "schemes"/protocols (project/path/myst). There is some redundancy here and we can possibly just simplify to `myst`, but we should make sure we don't loose anything. Alternatively we could go in the other direction, adding `download` instead of `path`?~ (removed)
- [x]  Some implications of the "scheme" choice are on downloads and how to download a markdown file easily, for example.
- [x] agree on syntax for interpolating text/title into a link template (currently suggested as `%t`, which is similar to `Fig. %s`). In sphinx this is `{name}`, which can also be supported, but maybe not mentioned in docs, etc. See @choldgraf response [here](https://github.com/executablebooks/myst-enhancement-proposals/issues/9#issuecomment-1376985293) - choosing `%t`.
- [x] ~update `python?py:class` --> `python:py:class`~ (removed intersphinx)

**Doc updates / TODOs**
- [x] Review "search order specificity" and "implicit" targets in a document
- [x] Add some information on the crossReference spec AST changes.
- [x] Add examples (I think these can be pretty light, as they are mostly laid out in the doc)

Current Status: `DRAFT`. Changing to `ACTIVE` needs to address above points, do a final pass, update the name of the doc and the ID.

We will aim to change this to active for review by wider stakeholders and review (aiming to take @mmcky through in part on Monday) in the next few days and start to get their review.

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,397 @@
+| `` {doc}`my-doc` ``                               | `[](../examples/my-doc.md)`                |
+| `` {download}`my-doc.zip` ``                      | `[](my-doc.zip)`                           |
+| `` {py:class}`zipfile.ZipFile` ``                 | `[](#zipfile.ZipFile)`                     |
+| `` {external+python:py:class}`zipfile.ZipFile` `` | `[](myst:python?py:class#zipfile.ZipFile)` |
```
In https://github.com/executablebooks/MyST-Parser/pull/673 I have changed this slightly (see also https://myst-parser.readthedocs.io/en/latest/syntax/syntax.html#cross-project-inventory-links):

1. `inv` instead of `myst`, to denote that it is referencing an `objects.inv` file
2. `python:py:class` instead of `python?py:class`, where e.g. `python:py` would be resolved as `python:py:*` and `python` as `python:*:*`
3. `*` is a wildcard that can match zero or more characters, and `\*` matches the literal `*`

rowanc1 replied:
The moving of `python?py:class` --> `python:py:class` makes sense to me, as well as the wildcard.

Note about `inv` in comments below.

rowanc1 replied:
I have removed all mentions of intersphinx now, and am opting to discuss this in a future MEP. This keeps this one smaller and focust on internal project cross references.

rowanc1 commented:
@chrisjsewell from a few comments you seemed to have strong opinions about the `myst` url scheme - do you think that is the only thing holding this PR from moving into `ACTIVE`?

My thoughts on that are in the future we want a scheme to do more than what `inv` does (e.g. enumeration and dynamic linking of _content_), this is a core part of what "structured" means to me in My**S**T, and allows for rich interlinking between projects in a way that goes beyond the `inv` file format.

That being said, I think there is a lot of great work in the thinking behind links and we are already adopting it in our various parsers.

**What I propose:**
1. change `myst` to `inv` in this iteration, this hopefully remove any sticking points of this MEP
2. agree that `myst:` scheme will be a scheme based on the myst-spec for resolving references, but that can be hashed out in a future MEP

If that sounds good, I can take a pass on cleaning up the final things in this PR and we can get it over to `ACTIVE` for a wider/final review.

chrisjsewell commented:
> @chrisjsewell<https://github.com/chrisjsewell> from a few comments you seemed to have strong opinions about the myst url scheme - do you think that is the only thing holding this PR from moving into ACTIVE?

Heya, ok let me have a think and try to get this sorted by the weekend 😅
As you’ve seen I’ve heard be been trying to push this forward in some sense in myst-parser. Obviously the big pain in the backside there is trying to transition in a « back-compatible » way.
In terms of `myst` yeh hmm, I see your opinion, but it’s also not very « descriptive » of what it actually is 🤔

Chris

On 10 Feb 2023, at 23:44, Rowan Cockett ***@***.***> wrote:

﻿

@chrisjsewell<https://github.com/chrisjsewell> from a few comments you seemed to have strong opinions about the myst url scheme - do you think that is the only thing holding this PR from moving into ACTIVE?

My thoughts on that are in the future we want a scheme to do more than what inv does (e.g. enumeration and dynamic linking of content), this is a core part of what "structured" means to me in MyST, and allows for rich interlinking between projects in a way that goes beyond the inv file format.

That being said, I think there is a lot of great work in the thinking behind links and we are already adopting it in our various parsers.

What I propose:

  1.  change myst to inv in this iteration, this hopefully remove any sticking points of this MEP
  2.  agree that myst: scheme will be a scheme based on the myst-spec for resolving references, but that can be hashed out in a future MEP

If that sounds good, I can take a pass on cleaning up the final things in this PR and we can get it over to ACTIVE for a wider/final review.

—
Reply to this email directly, view it on GitHub<https://github.com/executablebooks/myst-enhancement-proposals/pull/10#issuecomment-1426439120>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAW32QSCFOJQ4D6INBENTADWW3AE7ANCNFSM6AAAAAATXC7YPQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

rowanc1 commented:
Nice. All I am proposing at this stage is that we kick the conversation about the `myst:` scheme down the road a bit and focus on getting a smaller MEP out the door.

For the rest of the MEP: are there outstanding parts you know/suspect are not backwards compatible? Perhaps we could take a similar approach. That is, make the MEP smaller and focus on shipping it and starting to turn these links on by default in various places!

choldgraf commented:
FWIW, I'll leave it to you two to decide the best path forward, but I agree that smaller and more tightly scoped proposals are easier to explain, discuss, and decide on. We can always make future progress in a new iteration.

choldgraf commented:
FWIW, I'll leave it to you two to decide the best path forward, but I agree that smaller and more tightly scoped proposals are easier to explain, discuss, and decide on. We can always make future progress in a new iteration.

rowanc1 commented:
Ok, I have removed the following content, which is all things about intersphinx. I am opting to just kick that over to another MEP that can come in at any time. That is where most of the questions were, and I think the rest is actually already implemented in all parsers (maybe not released yet).

@chrisjsewell can you take a look soon, hopefully nothing left that is outstanding, then we can get this in and turn it to active and get a bit wider review.

````
| `` {py:class}`zipfile.ZipFile` ``                 | `[](#zipfile.ZipFile)`                     |
| `` {external+python:py:class}`zipfile.ZipFile` `` | `[](myst:python?py:class#zipfile.ZipFile)` |


- `` {external+python:py:class}`zipfile.ZipFile` `` - intersphinx
- Custom `extlinks` through simple [Sphinx configuration](https://www.sphinx-doc.org/en/master/usage/extensions/extlinks.html)

 Use the query string to filter matches by `domain:object_type`. Use `*` to match any value, e.g. `*:term` will match term in any domain. For example, `[text](?std:label#api/main)` searches only in the `std:label` domain.



Note that the `domain:object_type` syntax is specific to Sphinx, and may change or depend on the `key` that is being referenced. For example, targeting a [JATS document](https://jats.nlm.nih.gov/archiving/tag-library/1.3/attribute/ref-type.html), the reference kind can be `fig`, `sec`, `chem` etc. and doesn't have a concept of a domain. The `myst:` query syntax is designed to be flexible and extensible to these use cases.

For example, `[my custom text](myst:python#zipapp-specifying-the-interpreter)` becomes:

```yaml
- type: link
  url: https://docs.python.org/3.7/library/zipapp.html#zipapp-specifying-the-interpreter
  urlSource: myst:python#zipapp-specifying-the-interpreter
  children:
    - type: text
      value: my custom text
  data:
    title: Specifying the Interpreter
    inv: https://docs.python.org/3.7/objects.inv
    version: '3.7'
    enumerator: '2.3'
    lang: en
  internal: false
  protocol: myst
```

**InterSphinx Protocol**
: There was [discussion](https://github.com/executablebooks/MyST-Parser/pull/613#discussion_r968548359) on what protocol to use for cross-project links (we settled on `myst:`). Other possibilities included `inv:`, we opted for `myst` to reinforce the branding of the project and ensure the concept could be extended to other, richer, content links in the future that were not ".inv" files specifically.

````

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+| `` {ref}`my-id` ``                       | `[](#my-id)`                 |
+| `` {eq}`my-equation` ``                  | `[](#my-equation)`           |
+| `` {ref}`Custom Text <my-id>` ``         | `[Custom Text](#my-id)`      |
+| `` {ref}`See "{name}" <my-id>` ``        | `[See "%t"](#my-id)`         |
```
```suggestion
| `` {numref}`See "{name}" <my-id>` ``        | `[See "%t"](#my-id)`         |
```

rowanc1 replied:
done.

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+| `` {ref}`my-id` ``                       | `[](#my-id)`                 |
+| `` {eq}`my-equation` ``                  | `[](#my-equation)`           |
+| `` {ref}`Custom Text <my-id>` ``         | `[Custom Text](#my-id)`      |
+| `` {ref}`See "{name}" <my-id>` ``        | `[See "%t"](#my-id)`         |
```
```suggestion
| `` {numref}`See "{name}" <my-id>` ``        | `[See "%t"](#my-id)`         |
```

rowanc1 replied:
done.

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+
+## Summary
+
+We propose a cross-reference syntax that uses CommonMark links to support all use cases of cross-referencing content internal to a project. The syntax aims to be familiar and work across different rendering platforms. Most internal content can be referenced using a hash-link, `[](#my-id)`, which is the recommended replacement for the multiple role options that can do this in MyST currently (e.g.  ...
```
```suggestion
We propose a cross-reference syntax that uses CommonMark links to support all use cases of cross-referencing content internal to a project.
The syntax aims to be familiar and work across different rendering platforms.
Most internal content can be referenced using a hash-link, `[](#my-id)`, which is the recommended replacement for the multiple role options that can do this in MyST currently (e.g. `` {ref}`my-id` ``, `` {eq}`my-id` ``, `` {numref}`my-id` ``).
We provide options for increasing specificity for these links in all cases to deal with duplicate references across pages in a project.
```

rowanc1 replied:
👍

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+
+If the `text` is included it will be used as is with two additional template values (`%s` and `%t`)
+
+- Enumeration (`%s`)
```
I don't think `%s` makes sense here.
The fact that sphinx uses it (as an old style format), is purely just to comply with https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting
But here, if we are going to use `%t` for title, then enumeration should similarly be represented by `%e`

rowanc1 replied:
Maybe `%n`, I think that most people will think "number" rather than enumeration, even though that is what the spec is.

mmcky replied:
Does this have to be a single letter. Could we use `%num`? I understand the idea behind `%e` (relative to `%t`) but I honestly don't think I would remember to us it.

FWIW - As a user I tend to make use of the `numref` role and accept the default `style` of numbered figures. I don't -- but a user could change the template for the `numref`  format and specify something like `Fig %s` rather than displaying `Figure %s` (for example) to change this document wide right?

rowanc1 replied:
Yes! Regardless of the letter/sequence we use to interpolate the number into the text, it will be configurable at a project, page, and link/numref level.

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+  - Any reference target that is enumerated can reference that number or string with a `%s`.
+  - If the target is enumerated, the default will be the numbered form of the reference (e.g. "Section 2.1.2", "Fig. 3", or "(1)")
+  - If a `%s` is used and the node is not enumerated, the `%s` will be replaced by "??" and a warning raised.
+  - Parsers can optionally choose to support `{number}` which is from Sphinx.
```
I don't think this should be an option; we pick one format and stick with that, otherwise its an unnecessary complication for implementations and users

rowanc1 replied:
Done.

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+- Enumeration (`%s`)
+
+  - Any reference target that is enumerated can reference that number or string with a `%s`.
+  - If the target is enumerated, the default will be the numbered form of the reference (e.g. "Section 2.1.2", "Fig. 3", or "(1)")
```
So are we saying here that e.g. `%s == "Section 2.1.2"`, if so should there not be another placeholder for e.g. `%e == "2.1.2"`

rowanc1 replied:
I have clarified this to say:

> If the target is enumerated and there is no text provided by the link, the text will be the numbered form of the reference (e.g. "Section 2.1.2", "Fig. 3", or "(1)" depending on the node and parser options)

If you are over-riding the default reference style, then all you should need is the number, and if you want the reference to include "Figure 1", it isn't too much effort to add `[in Figure %s]` (with whatever the letter we choose is..).

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+
+### Search Order and Specificity
+
+All references search the local document first[^specific_doc], then the local project in the order of the table of contents. A `xref_multiple` warning is raised if multiple matches are found.
```
```suggestion
All references search the local document first[^specific_doc], then the local project in the order of the table of contents. A `xref_ambiguous` warning is raised if multiple matches are found.
```

this is already used in myst-parser

rowanc1 replied:
Nice, that is better naming.

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+
+[^specific_doc]: With the exception of an explicit reference to a specific page, i.e. `[](./examples/my-doc.md#explicit-reference)`
+
+In large documentation sites, a referenced target can be present in multiple documents, in that case, the parser will emit a `xref_multiple` warning letting you know that there are multiple matches for the intended target.
```
```suggestion
In large documentation sites, a referenced target can be present in multiple documents, in that case, the parser will emit a `xref_ambiguous` warning letting you know that there are multiple matches for the intended target.
```

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+- enforce uniqueness via suffix enumeration `-1`
+
+For example, `## Links and Referencing` can be referenced as `[](#links-and-referencing)`.
+These are **implicit** references, and referring to them should raise an `xref_implicit` warning, which can optionally be suppressed by users.
```
Note, in myst-parser, I currently don't emit this warning, because it is optional anyway to turn on anchors, to a certain depth.

Is this MEP suggesting every heading level should have an anchor, I think that should be said explicitly

Also this MEP does not mention about a `[](path/to/file.md#heading-anchor)` format.
This is something that GitHub, VS Code, etc support, as does Myst-Parser. Should it be here?

rowanc1 replied:
I have lightened the language, from `should` to `can`, I think that we will continue to raise that warning, so would like it documented, but also not a big deal as there are different ways to bring the configuration/parser together.

Added the text `Every heading level in a document should have an anchor`

The `[](path/to/file.md#heading-anchor)` is mentioned on around line 257 as "Target in a document", I didn't spend too long explaining it, but it is definitely there!

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+`xref_unsupported`
+: Raised if the the current environment does not support the reference look up. For example, single page builds.
+
+`xref_multiple`
```
```suggestion
`xref_ambiguous`
```

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+`xref_multiple`
+: Raised when multiple conflicting targets are matched.
+
+`xref_duplicate`
```
Does this need to be separate from the missing/ambiguos warning?

rowanc1 replied:
Removed.

chrisjsewell reviewed (changes requested):
See comments 😄

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+| Target in a document     | `<project:target.md#file>` | `[](file.md#target)`      |
+| Target in project        | `<project:#target>`        | `[](#target)`             |
+
+### Search Order and Specificity
```
One thing that is not mentioned here is the "fall back" behaviour, i.e. what happens if a link cannot be resolved.

In myst-parser I recently changed this from outputting nothing to outputting an external link, i.e. `<a href="#target">#target</a>`

rowanc1 replied:
Put in:

> If a link cannot be resolved, an external link should be rendered, for example, `<a href="#target">#target</a>`.

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+- If the reference target is not enumerated, the title will be used. This may be the section header text (including syntax/style) or a figure caption.
+- If the node is not enumerated and does not have a title, the reference label will be used.
+
+If the `text` is included it will be used as is with two additional template values (`%s` and `%t`)
```
A standard feature in CommonMark is https://spec.commonmark.org/0.30/#backslash-escapes

Should you also be able to escape a `%` in this context, to stop it from being replaced?

rowanc1 replied:
Added:

> In both cases, the `%` can be escaped with a preceding backslash, that is `\%s` or `\%t`, and the text will not be replaced.

fwkoch commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+| `` {ref}`Custom Text <my-id>` ``         | `[Custom Text](#my-id)`      |
+| `` {ref}`See "{name}" <my-id>` ``        | `[See "%t"](#my-id)`         |
+| `` {numref}`Custom Number %s <my-id>` `` | `[Custom Number %s](#my-id)` |
+| `` {doc}`my-doc` ``                      | `[](my-doc.md)`              |
```
In the new syntax, it looks like you always need to provide the extensions when referencing other pages? Is there any implicit resolution of `[](my-doc)`? I could see this getting confusing if source files have different extensions (`md`, `ipynb`, `tex`...), particularly in external projects. (Maybe that's getting into complexity with intersphinx, but still seems relevant within a single project.)

rowanc1 replied:
Resolved in a comment below!

fwkoch commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+
+- These follow the same path rules as above when referring to a unknown file-type (e.g. `./my-file.txt`).
+- Files with known file extensions (e.g. `*.md`, `*.ipynb`) will default to being document links, not downloads.
+- Downloads can be specified explicitly by a `<path:my-file.md>`, which will revert to a download regardless of the extension. (i.e. this overrides the default for `[](my-file.md)`, which is `<project:my-file.md>`).
```
You can also add the scheme in the `inline` link, right? i.e. `[](path:my-file.md)`

rowanc1 replied:
Yes, I will make that explicit!

rowanc1 replied:
I have updated teh text to:

> In most cases, as seen in the summary above the scheme is optional and can be inferred safely by the context.

And we have an example of that in one of the tables as `[](path:file.md)`.

fwkoch commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+- Follow web-standards / conventions for URLs where possible (e.g. query strings, protocols)
+- Design new additions to the `myst-spec` AST that can provide rich information to renderers
+
+These link improvements are completed in the context of supporting (1) academic citations; and (2) intersphinx cross-references; however, this MEP does not specifically support the intricacies of intersphinx, bibliographies or referencing. We encourage a future MEPs to address these concerns.
```
```suggestion
These link improvements are completed in the context of supporting (1) academic citations; and (2) intersphinx cross-references; however, this MEP does not specifically support the intricacies of intersphinx, bibliographies or referencing. We encourage future MEPs to address these concerns.
```

rowanc1 replied:
Thank you!

fwkoch commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+- Links can be absolute paths from the project root.
+- Absolute paths start with `/`.
+- The path separator is POSIX `/`.
+- The path must include the extension.
```
Ah, I think this answers my previous question: no implicit resolution of `[](my-doc)` or `[](../supplementary/my-doc)` without extensions.

rowanc1 replied:
Yeah, I think this is better and gets rid of ambiguity!!

fwkoch commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+
+#### `link`
+
+The links are defined by a scheme, which can be standard protocols (`http:`, `mailto:`). Here we propose three new schemes, `path`, `project`, and `myst`, which is an extensibility point described by [CommonMark](https://spec.commonmark.org/0.30/#example-598). These schemes are used to indicate that the link should be resolved by MyST specific logic, and follows standard [URI][uri] syntax:
```
Should we even mention the `myst` scheme here? It is not explained anywhere else in this document, and maybe gets more into future MEP intersphinx territory

rowanc1 replied:
Thank you, didn't catch this one on my first pass, this is now removed.

rowanc1 commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,339 @@
+
+## Summary
+
+We propose a cross-reference syntax that uses CommonMark links to support all use cases of cross-referencing content internal to a project. The syntax aims to be familiar and work across different rendering platforms. Most internal content can be referenced using a hash-link, `[](#my-id)`, which is the recommended replacement for the multiple role options that can do this in MyST currently (e.g.  ...
```
Updated.

chrisjsewell commented:
> Ok, incorporated all of the reviews except the `%s` or `%e` thing for the number.
> I think leaning on that rather than introducing another thing to remember is better.


But what does `s` signify?
In the MEP you mention "string", but then what is the title if its not a string?
I think `%s` vs `%t` is confusing; one your referring to by the "type" of the variable, the other you are referring to by the "name" of the variable

rowanc1 commented:
> But what does `s` signify?
Nothing. This isn't new behaviour it is what Sphinx/RST/MyST already supports for the `numref` role, and is just documenting it here. The current behaviour for `numref` is documented here:
https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-numref

As an alternative to move this forward, we could say in the enumeration section:

> We continue to use and support the enumeration and name interpolation of the `{numref}` role, which is now also available in the markdown link.

And also paraphrase the existing behaviour.

chrisjsewell commented:
> The current behaviour for numref is documented here:

Sphinx specifically marks in the code that this is the "old style" (i.e. legacy) format: https://github.com/sphinx-doc/sphinx/blob/30f347226fa4ccf49b3f7ef860fd962af5bdf4f0/sphinx/domains/std.py#L902

I just don't feel we should base our solution on something that sphinx has already deemed to be subpar, and is essentially only supporting for back-compatibility

I also don't think, for a "new" part of the spec, we should be already be fragmenting into different ways to do the same thing. We should really just pick one and go with that.

**But basically; if you change `%s` to `%e`, then I would sign off on this**

---

I would also highlight a key difference with Sphinx `numref`; there the `text` is just that (plain text), whereas in MyST `[text](link)` is not actually text, its inline syntax, which makes things a bit more "fiddly".
With these placeholders, I do want to stress, you are introducing a new, non-CommonMark, syntax token. We should be very careful about that.

---

After talking here, I actually feel using **only** the `{name}` and `{number}` placeholders is perhaps better (or even `{title}`and `{enum}`);
its a lot more intuitive what it is, without having to read the spec:
`[{number}](#link)` vs `[%s](#link)`

this could also be extensible, if we / other implementations wanted to add additional placeholders

The key hesitation I have though, is that braces `{}` are already used for roles/directives (and maybe now attributes)

choldgraf commented:
I think that if we can't get to consensus on this quickly, then we should just remove the "title and number interpolation syntax" aspect of the MEP, and leave it to implementations to define this (and perhaps use numref as an example and note well visit this in the future). We can make this a targeted MEP in the future if it's important to standardize. If y'all think it's crucial to block the MEP on this topic then that's fine too, but IMO it doesn't feel like something worth blocking on

rowanc1 commented:
> I actually feel using **only** the `{name}` and `{number}` placeholders is perhaps better

I am good with that, this is exactly the same as the "new" sphinx syntax, so we aren't introducing anything new here.
I will update the text accordingly.

chrisjsewell commented:
> I am good with that, this is exactly the same as the "new" sphinx syntax, so we aren't introducing anything new here.
I will update the text accordingly.

Thanks @rowanc1.
As I mentioned the only thing I'm a little worried about is the "proliferation" of braces lol.

For example, you could theoretically get something like this `` [{name} {name}`this is actually a role`{.a c="this is an inline attribute"}](#link) ``

Do you think thats not a bit eurgh?

Also to clarify (and maybe update this in the mep), does `{number}` go to e.g. `Section 2.1.2` or `2.1.2`?
and for whatever one it is not, should there be another placeholder, that can give you it?

rowanc1 commented:
I have clarified that `{number}` is just the enumerator. I don't think that we need one for interpolating the default "Figure 1", but we could certainly add that down the line if there is demand.

I agree that all the brackets that could be get visually messy, but I don't think that it will be common in practice. Most of the time people will use the default auto-reference (i.e. if coming from `\autoref` in latex). In my experience when people over-ride that default, I haven't ever come across anything complicated (e.g. "see here" and "in figure {number}" are common, sometimes with some bold text).

chrisjsewell commented:
> I agree that all the brackets that could be get visually messy, but I don't think that it will be common in practice.

yep it is not the end of the world, but something I did want to emphasize.

> (i.e. if coming from` \autoref` in latex

I always used http://tug.ctan.org/tex-archive/macros/latex/contrib/cleveref/cleveref.pdf 😉 , shame we can't do reference "concatenation", but let's not get ahead of ourselves

Almost there although I just noticed something....

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,360 @@
+| `` {ref}`my-id` ``                       | `[](#my-id)`                 |
+| `` {eq}`my-equation` ``                  | `[](#my-equation)`           |
+| `` {ref}`Custom Text <my-id>` ``         | `[Custom Text](#my-id)`      |
+| `` {numref}`See "{name}" <my-id>` ``     | `[See "%t"](#my-id)`         |
+| `` {numref}`Custom Number %s <my-id>` `` | `[Custom Number %s](#my-id)` |
```
```suggestion
| `` {numref}`See "{name}" <my-id>` ``     | `[See "{name}"](#my-id)`         |
| `` {numref}`Custom Number {number} <my-id>` `` | `[Custom Number {number}](#my-id)` |
```

rowanc1 replied:
Good catch. Updating.

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,360 @@
+| `` {ref}`my-id` ``                       | `[](#my-id)`                 |
+| `` {eq}`my-equation` ``                  | `[](#my-equation)`           |
+| `` {ref}`Custom Text <my-id>` ``         | `[Custom Text](#my-id)`      |
+| `` {numref}`See "{name}" <my-id>` ``     | `[See "%t"](#my-id)`         |
+| `` {numref}`Custom Number %s <my-id>` `` | `[Custom Number %s](#my-id)` |
```
```suggestion
| `` {numref}`See "{name}" <my-id>` ``     | `[See "{name}"](#my-id)`         |
| `` {numref}`Custom Number  {number} <my-id>` `` | `[Custom Number {number}](#my-id)` |
```

chrisjsewell commented on meps/mep-cross-references.md:
````diff
@@ -0,0 +1,360 @@
+url.hash; // "#my-ref"
+```
+
+For most internally linked references, we expect the inline syntax to be most commonly used, with the autolink and scheme to only be used when specific behavior is intended. The following links and references are supported:
````
It is of note that one key use-case for auto-links, that has been requested a number of times, is that they integrate better with non-myst renderers, i.e. here `[](#target)` will show (nothing) [](#target), but  `<project:#target>` will show  <project:#target>

chrisjsewell replied:
I was going to suggest also having something like `[{default}](#link)`, but I guess the auto-link covers it?

rowanc1 replied:
I have removed all but the end of that line:

> The following links and references are supported:

I think auto link does cover the case you are referring to.

rowanc1 replied:
Marking this thread as resolved, I think that the changes fixed this up?

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,360 @@
+Files that are outside of the table of contents of the project and are referenced directly are downloads.
+
+- These follow the same path rules as above when referring to a unknown file-type (e.g. `./my-file.txt`).
+- Files with known file extensions (e.g. `*.md`, `*.ipynb`) will default to being document links, not downloads.
```
maybe to note, their implicit text will be that of the page title/top-header

rowanc1 replied:
Good call. Added:

> , with the document title being used as the default text.

chrisjsewell commented on meps/mep-cross-references.md:
```diff
@@ -0,0 +1,360 @@
+
+- These follow the same path rules as above when referring to a unknown file-type (e.g. `./my-file.txt`).
+- Files with known file extensions (e.g. `*.md`, `*.ipynb`) will default to being document links, not downloads.
+- Downloads can be specified explicitly by a `<path:my-file.md>`, which will revert to a download regardless of the extension. (i.e. this overrides the default for `[](my-file.md)`, which is `<project:my-file.md>`).
```
An implementation note: `path` was chosen over, the perhaps more obvious `file`, because markdown-it (and possibly other parsers?) explicitly block `file` schema, to prevent XSS attacks

chrisjsewell replied:
oops no I see this down in the bottom section

chrisjsewell reviewed (changes requested):
just a few little things

rowanc1 commented:
I think those are all of the comments @chrisjsewell and @fwkoch, please reopen or add any more if you find other things.

chrisjsewell reviewed (approved):
All good cheers

rowanc1 commented:
@choldgraf @chrisjsewell @fwkoch I have updated and rebased the MEP, and changed the authorship to alphabetical, first name order. I have set the date for tomorrow, added the `0002` id, and changed the status to `Active`.

@choldgraf I think this is ready to (squash) merge! 🎉🎉🎉🎉 

And then we can [Step 4: Activate decision making](https://myst-eps.readthedocs.io/en/latest/overview/#step-4-activate-decision-making)!

🚀

chrisjsewell commented:
@choldgraf I think either you want to add back the `pre-commit-config.yml`, or deactivate the pre-commit on this repo

colliand commented:
Under link: "Here we propose three new schemes, path and project which is an extensibility point described by CommonMark." I think that's only two schemes not three. 

As a Markdown user who wants to use Myst more, I like the proposed syntax changes!

agoose77 commented:
I'm very interested in this new syntax!

Regarding the templating syntax `{number}` in `[Custom Number {number}](#my-id)`, are we precluding the ability to embed roles in the text of a link? Or do we now reserve `{number}` and `{name}` as identifiers that are not available to roles? I am thinking about the case that users want to craft interesting links that contain e.g. maths, it might look like
```
[If {math}`x` is an integer, the go to {name}](#x-is-integral)
```

If we're happy with this, fine, just wanted to be sure we've discussed it!

chrisjsewell commented:
> are we precluding the ability to embed roles in the text of a link

`{xxx}` is not parsed as a role, it has to be attached to a literal

agoose77 commented:
> `{xxx}` is not parsed as a role, it has to be attached to a literal

Fab. I think what I'm trying to articulate is a question as to whether using the same tokens (in a different context) will cause problems. If not, and we can parse the full inline set of constructs, then I'm stoked

chrisjsewell commented:
> > `{xxx}` is not parsed as a role, it has to be attached to a literal
> 
> 
> 
> Fab. I think what I'm trying to articulate is a question as to whether using the same tokens (in a different context) will cause problems. If not, and we can parse the full inline set of constructs, then I'm stoked

https://github.com/executablebooks/myst-enhancement-proposals/pull/10#issuecomment-1446703798

agoose77 commented:
OK, good. I hadn't appreciated that Sphinx is also moving to a new syntax. I'm glad we're on the same page. Visually, I like it.

chrisjsewell commented:
> OK, good. I hadn't appreciated that Sphinx is also moving to a new syntax. I'm glad we're on the same page. Visually, I like it.

Yep 👍 this MEP derives from https://github.com/executablebooks/MyST-Parser/pull/613

rowanc1 commented:
@agoose77 

Can see a demo of [what you are describing here](https://myst-tools.org/sandbox?tab=demo&myst=JAAkACAAXABpAG4AdABfAHsAYQB9AF4AewBiAH0AIAB4AF4AMgAgAFwALABkAHgAIAAkACQAIAAoAHgALQBpAHMALQBpAG4AdABlAGcAcgBhAGwAKQAKAAoAWwBJAGYAIAB7AG0AYQB0AGgAfQBgAHgAYAAgAGkAcwAgAGEAbgAgAGkAbgB0AGUAZwBlAHIALAAgAHQAaABlAG4AIABnAG8AIAB0AG8AIAB7AG4AdQBtAGIAZQByAH0AXQAoACMAeAAtAGkAcwAtAGkAbgB0AGUAZwByAGEAbAApAA%3D%3D), it is actually really nice. :)

@colliand, thanks for the catch on the numbering, I have fixed the text.

fperez commented:
I only wanted to thank you all for the team and express my general happiness with these proposed changes!! I have no specific concerns to raise, though I admit I didn't dig into all the details. But the goals set at the top are excellent ones to optimize for, and every detail I did check looked great to me. 🚀

agoose77 commented:
The choice of URI schemes also looks reasonable; they're not already-reserved identifiers in the IANA database, that I can see.

agoose77 commented:
I notice in the rendered document that a `note` admonition isn't rendering correctly. The syntax looks correct at first glance, any ideas?

![image](https://user-images.githubusercontent.com/1248413/222906220-b204f092-93e5-4474-8100-f2643391fa6d.png)

chrisjsewell commented:
> I notice in the rendered document that a note admonition isn't rendering correctly. The syntax looks correct at first glance, any ideas?

fix in #19; It's not "core" MyST yet 😉

stevejpurves commented:
A few comments and questions from coming into this fresh, nothing to stop this going ahead as is though:

* the new syntax is very nice and simplifies things enormously. Seems easier to reason about overall even though various subtleties are being hidden from the user
* the range of things you can do inside the link is really powerful - including the string interpolation on `{number}` and `{name}`. The syntax choice there is natural as well, especially for python users
* Q: I understand the introduction of two new schemes -- but from the MEP I'm not clear what they are for other than signally internal myst logic should handle those schemes -- I understand `project:` will conduct a search over the project files in the toc, does this mean that `path` is only ever used in the case of downloads? in particular it takes a while to understand that  use of `path:` in the "File Download (explicit)" in the table [here](https://myst-eps--10.org.readthedocs.build/en/10/meps/mep-0002/#link) - is just to explicitly remove the ambiguity between that and the the "Project Document" example. i.e. if the file is in the TOC but we want to have a download link then the `path:` scheme is included? is that correct?
* will the current MyST syntax (i.e. `[](my-label)`) be deprecated in favor of `[](#my-label)`? is that what issuing the `xref_legacy` warning achieves/signals?

Overall though this is a great enhancement and a well thought though MEP 👏🏼 - definitely I don't want to be writing or thinking about any of the other cross reference variants - `[]()` is the way!

stevejpurves commented on docs/meps/mep-0002.md:
```diff
@@ -0,0 +1,364 @@
+| External URL             | `<https://example.com>`    | `[](https://example.com)` |
+| Local file download      | `<path:file.txt>`          | `[](file.txt)`            |
+| File download (explicit) | `<path:file.md>`           | `[](path:file.md)`        |
+| Project document         | `<project:file.md>`        | `[](file.md)`             |
```
from the table here I found some of the ambiguity confusing. The syntax is fine though, it was just the way it was expressed in the MEP -- but reading on, the Downloads section below does make it clearer what is going on between  `[](path:file.md)` and `[](file.md)` which is (i) download file that is also in the TOC and (ii) link to a file in the TOC(?)

So I think that is fine.

chrisjsewell commented:
> will the current MyST syntax (i.e. [](my-label)) be deprecated in favor of [](#my-label)? is that what issuing the xref_legacy warning achieves/signals?

yes

fwkoch commented on docs/meps/mep-0002.md:
```diff
@@ -0,0 +1,364 @@
+The links should follow the [link AST](https://www.myst.tools/docs/spec/myst-schema#link) for external links.
+For internal project cross-references, these should be resolved to a `crossReference` node ([spec](https://www.myst.tools/docs/spec/myst-schema#crossreference)).
+
+For external project links, these extend the link object with additional data that includes the url source (`urlSource`), the scheme name (e.g. `project` or `download`), whether the link is internal (e.g. `false`), and additional optional metadata about the page that may be helpful to a renderer.
```
```suggestion
For external project links, we will extend the link object with additional data that includes the url source (`urlSource`), the scheme name (e.g. `project` or `download`), whether the link is internal (e.g. `false`), and additional optional metadata about the page that may be helpful to a renderer.
A prototype of this extended link type is currently defined in [myst-transforms](https://github.com/executablebooks/mystjs/blob/327c19c75c2f743e9f0736586b5232a45e00123e/packages/myst-transforms/src/links/types.ts#L4-L10).
```

fwkoch reviewed (approved):
The changes since my last review look good!

Only one small comment in the `Specification AST` section to make it slightly more clear we are changing the existing `myst-spec` `link` (and a pointer to where we have already implemented the changes).

chrisjsewell commented on docs/meps/mep-0002.md:
```diff
@@ -324,7 +324,8 @@ Files that are outside of the table of contents of the project and are reference
 
-For external project links, these extend the link object with additional data that includes the url source (`urlSource`), the scheme name (e.g. `project` or `download`), whether the link is internal (e.g. `false`), and additional optional metadata about the page that may be helpful to a renderer.
+For external project links, we will extend the link object with additional data that includes the url source (`urlSource`), the scheme name (e.g. `project` or `download`), whether the link is internal (e.g. `false`), and additional optional metadata about the page that may be helpful to a renderer.
+A prototype of this extended link type is currently defined in [myst-transforms](https://github.com/executablebooks/mystjs/blob/327c19c75c2f743e9f0736586b5232a45e00123e/packages/myst-transforms/src/links/types.ts#L4-L10).
```
I would rather this went in the reference section, since it is mentioning an implementation and not a spec. Bit it’s not the end of the world

chrisjsewell replied:
It’s probably better to just copy the actual « spec proposal » here

rowanc1 replied:
Yeah, agreed. Moved to the references!

gregcaporaso reviewed (approved):
This all sounds great to me. I think the new syntax will be extremely convenient.

fperez reviewed (approved):
I have two minor suggestions - I'd appreciate the 2nd in particular being addressed, but don't consider ti a blocker (the first is trivial).

rowanc1 commented on docs/meps/mep-0002.md:
```diff
@@ -0,0 +1,365 @@
+
+### Implicit Section Headers
+
+We suggest a configuration option to create anchor "slugs" for section headers, which stay close to the [GitHub implementation](https://github.com/Flet/github-slugger), which produces references that:
```
Done!

rowanc1 commented on docs/meps/mep-0002.md:
```diff
@@ -0,0 +1,365 @@
+## UX implications & migration
+
+All of the syntax is CommonMark compliant and introduces new capabilities to resolve cross references.
+All existing roles are being maintained for the forseeable future.
```
I have changed "foreseeable future" --> "indefinitely" and highlighted the compatibility with existing content and Sphinx as the reasons there.

The goal of this MEP is to simplify the amount of things that a user needs to learn before they are up and running with all of the cross-references, and it is designed to not break backwards compatibility with Sphinx roles. To that end I think leaving discretion up to various packages (e.g. jupyter-book, mystjs, myst-parser) on how they introduce these concepts is OK for now (e.g. implementations based on Sphinx may want to document things differently for their users who are familiar with RST).

It is a bit weaker than what you suggest @fperez, but that is probably ok for our first MEP. :)

Updated text:

> All existing roles are being maintained indefinitely to ensure compatibility with existing content as well as long-term compatibility with Sphinx. We suggest that documentation is updated to highlight the new, consistent markdown-link references with the old styles being moved to compatibility sections.

fperez replied:
Perfect, thanks!!

fperez commented:
All good from my end now, thanks! 🚀

rowanc1 commented:
Thank you @fwkoch @fperez @gregcaporaso @stevejpurves @chrisjsewell for the reviews! 🚀 I have brought all of the changes in that you suggested.

@stevejpurves I have updated the first time we introduced project/path schemes with a bit of explanatory text, I don't think that this materially changes anything, but does give the reader of the MEP a bit more of an idea of what is to come. @chrisjsewell @fwkoch can you review the change to make sure if it consistent please (added the middle few lines)!

> The links are defined by a scheme, which can be standard protocols (`http:`, `mailto:`).
Here we propose two new schemes, `project` and `path`:
the `project` scheme allows for cross-referencing pages, sections, equations, figures, or other components of a MyST project;
the `path` scheme allows for referencing files outside of the project or explicitly downloading the source of a document in the project.
The schemes are an extensibility point specifically described by [CommonMark](https://spec.commonmark.org/0.30/#example-598)
and are used to indicate that the link should be resolved by MyST specific logic, and follows standard [URI](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier) syntax

I am hoping that this high-level clarification also helps in some of the challenges you had @stevejpurves with understanding the [downloads section](https://myst-eps--10.org.readthedocs.build/en/10/meps/mep-0002/#downloads). Please let me know if that is the case!

fwkoch commented on docs/meps/mep-0002.md:
```diff
@@ -0,0 +1,368 @@
+the `project` scheme allows for cross-referencing pages, sections, equations, figures, or other components of a MyST project;
+the `path` scheme allows for referencing files outside of the project or explicitly downloading the source of a document in the project.
+The schemes are an extensibility point specifically described by [CommonMark](https://spec.commonmark.org/0.30/#example-598)
+and are used to indicate that the link should be resolved by MyST specific logic, and follows standard [URI](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier) syntax:
```
Just a tiny grammar comment, otherwise the new text seems good.

```suggestion
and are used to indicate that the link should be resolved by MyST specific logic. They follow standard [URI](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier) syntax:
```

rowanc1 replied:
Thanks. :)

chrisjsewell commented:
> can you review the change to make sure it is consistent please

yep all good cheers 👍

rowanc1 commented:
Awesome, thanks @chrisjsewell @fwkoch and @jstac as well!

I think that we are meeting all criteria from the MEP process for this being accepted! 🎉

📆 **MEP will close Monday March 13th, PT** 📆

As this is our first MEP running through the process, I propose we leave this open until **Monday March 13th, PT** (i.e. an additional five working days) and then if no other @executablebooks/core-team members ask for changes in that time, this is ✅✅✅✅✅✅ `ACCEPTED`.

For reference:
> A MEP may be accepted when all of the following conditions are met:
> - More than five (5) weekdays have passed since the proposal was marked as Active.
> - At least two PR Approvals from core team members.
> - No Request Changes from a core team member.

Thanks again to every one participating as we bootstrap this process. 🚀

rowanc1 commented:
Hi all - a bit delayed on hitting accept on this MEP, but that did give another 4 days beyond the deadline to have any other feedback. There hasn't been any, and we have the required conditions to merge the MEP!

I am going to squash merge this in now, and will coordinate a plan to update myst-spec over the next few weeks to reflect this MEP!

Thank you again for everyone participating in this process and helping bootstrap these MEPs into existence, this is the first of many enhancements that will improve MyST over the coming years. 🚀

choldgraf commented:
Wooo! Thanks Rowan for all of your work spearheading this one, and to everybody else for writing, commenting, discussing, etc.
